### PR TITLE
Fix double free error caused by failure of repalloc

### PR DESCRIPTION
**Pull request by pengzhout**
For big chunks larger than allocChunkLimit, gpdb call gp_realloc() to reallocate memory. Before reallocation, AllocFreeInfo() is called to decouple the chunk from shared header list, MemoryAcount system and chunk list, but gp_realloc() has a chance to fail due to OOM or low gp_vmem_protect_limit, in such cases, we need to put the chunk back to a health context to avoid double free and assertion.

### CODE REVIEW

**Comment by pengzhout** [comment]
You can set gp_vmem_protect_limit to 20 and write a UDF to reproduce this issue.
``` 
    size_t size = 10 * 1000;
    char * tmp = palloc0(size);
    PG_TRY();
    {
        size = 100 * 1000 * 1000;
        tmp = repalloc(tmp, size);
    }
    PG_CATCH();
    {
        pfree(tmp);
        PG_RE_THROW();
    }
    PG_END_TRY();
```

**Comment by yydzero** [comment]
As we have repo, could we add one test case into ICG?

File: src/backend/utils/mmgr/aset.c [hunk 1]
```diff
@@ -1638,16 +1635,23 @@ AllocSetRealloc(MemoryContext context, void *pointer, Size size)
 		Assert(block->freeptr == ((char *) block) +
 			   (chunk->size + ALLOC_BLOCKHDRSZ + ALLOC_CHUNKHDRSZ));
 
+		/* isHeader is set to false as we should never require realloc for shared header */
```
Comment (asimrp) [inline src/backend/utils/mmgr/aset.c]: Can you please shed some light on why the call to `AllocFreeInfo()` is moved from before the while loop to here?

Comment (pengzhout) [inline src/backend/utils/mmgr/aset.c]: As same as gp_realloc, it has a chance to report a error after the while loop, so the logic is do not free info before we found a block.

**Comment by pengzhout** [comment]
@foyzur can you help to review this since you also discovered this issue before.

**Comment by foyzur** [comment]
LGTM.

**Comment by pengzhout** [comment]
Thanks for the review, merged

**Pull request closed without merging**